src/Spice86.Core/Emulator/OperatingSystem/Dos.cs

```diff
@@ -197,11 +230,9 @@ public void AddDevice(IVirtualDevice device, ushort? segment = null, ushort? off
         _memory.UInt16[device.Segment, index] = device.InterruptEntryPoint;
         index += 2;
         if (device.Attributes.HasFlag(DeviceAttributes.Character)) {
-            var characterDevice = (CharacterDevice)device;
             _memory.LoadData(MemoryUtils.ToPhysicalAddress(device.Segment, index),
-                Encoding.ASCII.GetBytes( $"{characterDevice.Name,-8}"));
-        } else {
-            var blockDevice = (BlockDevice)device;
+                Encoding.ASCII.GetBytes( $"{device.Name,-8}"));
+        } else if(device is BlockDevice blockDevice) {
             _memory.UInt8[device.Segment, index] = blockDevice.UnitCount;
             index++;
```

Comment:
Not your code, but should really be done with a memory based structure, would make it much more obvious what is being done here.

Reply:
Done. Partially, at least.